src/lavinmq/mfile.cr

```diff
@@ -100,13 +103,30 @@ class MFile < IO
     self
   end
 
+  def reserve
+    @wg.add
+  end
+
+  def unreserve
+    @wg.done
+  end
+
   # The file will be truncated to the current position unless readonly or deleted
   def close(truncate_to_size = true)
-    # unmap occurs on finalize
     if truncate_to_size && !@readonly && !@deleted && @fd > 0
       code = LibC.ftruncate(@fd, @size)
       raise File::Error.from_errno("Error truncating file", file: @path) if code < 0
     end
+
+    # unmap if non has reserved the file, race condition prone?
+    if @wg.@counter.get(:acquire).zero?
```

Comment:
Since we're in `#close` I guess we can be sure here that counter won't be increased, only decreased? So no risk that counter is zero in this check but non-zero when unmap is called?

What about always spawning instead of this if-statement? It would be nice to not have to access an instance variable. Or maybe monkey patch WaitGroup with a `done?` method.

Reply:
"risk" that it is zero? the optimization here was for that fact that it will always be zero for single node clusters and for transient queues etc, and then we don't need to spawn a fiber. 

could monkey patch though.

Reply:
I mean if it's zero when we do the check, but it has changed when we call unmap so we're unmapping even though it's reserved. However, I think we can assume that the counter will never be increased at this point since we're in `#close` so the `MFile` shouldn't be used anymore?